### PR TITLE
(add) SONIC decoder

### PR DESCRIPTION
**Pull request by nepyope**
This PR ports SONIC's decoder network into lerobot. We showcase the possibilities that the integration unlocks by  replicating the [OpenHLM](https://huggingface.co/OpenHLM/OpenHLM-ckpts) paper: the authors finetune `openpi`'s `pi05` on trajectories recorded with SONIC teleop, we deviate from this strategy in that we train `pi05` directly on the `latent action token`, a 64-D vector that represents the robot's commanded trajectory in space.  This branch superseeds [#3827](https://github.com/huggingface/lerobot/pull/3827) focusing only on the releavant features. 

We introduce: 

## `SonicWholeBodyController`

A pure-Python/ONNX re-implementation of the **decode** half of NVIDIA's SONIC deploy stack for the Unitree G1. The encoder is intentionally absent: a token-output VLA (e.g. [`nepyope/sonic_walk`](https://huggingface.co/nepyope/sonic_walk)) supplies the 64-D latent `motion_token` directly each tick, and the SONIC **decoder** maps `token + recent proprioception history` to a residual action that is scaled and added onto the standing pose (`default_angles`) to produce 50 Hz joint-position targets for the robot's PD controller.

# Test

We teleoperate the real Unitree G1 through NVIDIA's `gear_sonic` SONIC whole-body controller. The SONIC encoder turns this teleop into the 64-D motion tokens, and the dataset is exported natively in LeRobot v2.1 format. This gives us 100 episodes of the ["walking back and forth" task](https://huggingface.co/datasets/nepyope/walk_back_and_forth). After converting the dataset to LeRobot v3, we then finetune [pi05](https://huggingface.co/lerobot/pi05_base).

- **Dataset:** [nepyope/walk_back_and_forth](https://huggingface.co/datasets/nepyope/walk_back_and_forth) (100 episodes)
- **W&B run:** [tino/lerobot/4h48jr8f](https://wandb.ai/tino/lerobot/runs/4h48jr8f?nw=nwusertino)
- **Trained model:** [nepyope/sonic_walk](https://huggingface.co/nepyope/sonic_walk)

## Training configuration

Following the recipe from the [OpenHLM paper](https://huggingface.co/nepyope/sonic_walk), we finetune `pi05_base` with a **cosine decay schedule with warmup**: the learning rate ramps linearly from `0` to a peak of `1e-4` over the first `500` warmup steps, then decays following a cosine curve down to `1e-5` over `5000` steps. We use the AdamW optimizer with `betas=[0.9, 0.95]` and a weight decay of `1e-4`, matching the paper's settings for stable large-model finetuning.

```bash
python -m lerobot.scripts.lerobot_train \
  --policy.type=pi05 \
  --policy.pretrained_path=lerobot/pi05_base \
  --policy.max_state_dim=64 \
  --policy.max_action_dim=64 \
  --policy.train_expert_only=true \
  --policy.freeze_vision_encoder=false \
  --policy.gradient_checkpointing=true \
  --policy.push_to_hub=true \
  --policy.repo_id=nepyope/sonic_walk \
  --dataset.repo_id=nepyope/walk_back_and_forth \
  --dataset.root=/fsx/martino/datasets/walk_back_and_forth_pi05 \
  --batch_size=32 \
  --num_workers=32 \
  --steps=5000 \
  --use_policy_training_preset=false \
  --optimizer.type=adamw \
  --optimizer.lr=1e-4 \
  --optimizer.weight_decay=1e-4 \
  --optimizer.betas=[0.9,0.95] \
  --scheduler.type=cosine_decay_with_warmup \
  --scheduler.num_warmup_steps=500 \
  --scheduler.num_decay_steps=5000 \
  --scheduler.peak_lr=1e-4 \
  --scheduler.decay_lr=1e-5 \
  --policy.chunk_size=50 \
  --policy.n_action_steps=50 \
  --wandb.enable=true \
  --wandb.disable_artifact=true \
  --save_freq=1000 \
  --output_dir=outputs/walk_back_and_forth_pi05 \
  --job_name=walk_back_and_forth_pi05
```

## Results

https://github.com/user-attachments/assets/1c1dc570-0ea3-4043-a7e8-0cba70e14fcf

The robot completes the task successfully. When placed out of distribution it still generalizes to the walking behavior.

> **TODO:** inspect pi05 internals to figure out how this generalization happens.

## Related PRs/branches

- **`unitree_locomotion`**: this is actually the gripper PR. Adds support for the `openarm` grippers that can be seen in the test video. 
- [**`g1_onboard_controller`**](https://github.com/huggingface/lerobot/pull/4267): forces the controller to run on the real robot.
- [**`onnx_helper`**](https://github.com/huggingface/lerobot/pull/4264):  fixes sim lag issues.
- [**`feat/sonic_motor_gains`**](https://github.com/huggingface/lerobot/pull/4262):  derive gains explicitly from `natural_freq` and `motor_model`
- **`openhlm`**:  fully faithful OpenHLM replication. Depends on the  **`sonic_encoder_decoder`** PR

### CODE REVIEW

**Comment by imstevenpmwork** [comment]
@claude Inspect the latest failing CI run for this PR. Read the failed job logs, identify the root cause, relevant error, and recommend a fix

File: docs/source/unitree_g1.mdx [hunk 1]
```diff
@@ -290,6 +290,47 @@ lerobot-rollout \
...
+lerobot-rollout \
+  --strategy.type=base \
+  --policy.path=nepyope/sonic_walk \
+  --policy.device=cuda \
+  --robot.type=unitree_g1 \
+  --robot.is_simulation=false \
+  --robot.robot_ip=<ROBOT_IP> \
+  --robot.controller=SonicWholeBodyController \
+  --robot.cameras='{"ego_view": {"type": "zmq", "server_address": "<ROBOT_IP>", "port": 5555, "camera_name": "head_camera", "width": 640, "height": 480, "fps": 30, "warmup_s": 5}}'
+```
```
Comment (haixuanTao) [inline docs/source/unitree_g1.mdx:359]: Should we add a part about running the controller onboard? and or run this very process onboard of the orin?

It could also be interesting to run the rollout without the g1 server no?

Comment (nepyope) [reply docs/source/unitree_g1.mdx:359]: https://github.com/huggingface/lerobot/pull/4267

i have a separate PR for the onboard controller, still in draft. You do need some sort of server for rollout since you can only connect to the robot via LAN.

**Comment by CarolinePascal** [comment]
@claude Review this PR

File: src/lerobot/robots/unitree_g1/controllers/__init__.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env python
+
+# Copyright 2025 The HuggingFace Inc. team. All rights reserved.
```
Comment (CarolinePascal) [inline src/lerobot/robots/unitree_g1/controllers/__init__.py]:
```suggestion
# Copyright 2026 The HuggingFace Inc. team. All rights reserved.
```

File: src/lerobot/robots/unitree_g1/controllers/sonic_whole_body.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
+#!/usr/bin/env python
+
+# Copyright 2025 The HuggingFace Inc. team. All rights reserved.
```
Comment (CarolinePascal) [inline src/lerobot/robots/unitree_g1/controllers/sonic_whole_body.py]:
```suggestion
# Copyright 2026 The HuggingFace Inc. team. All rights reserved.
```

File: src/lerobot/robots/unitree_g1/controllers/sonic_whole_body.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        self.action_ft = {f"{TOKEN_ACTION_PREFIX}.{i}.pos": float for i in range(TOKEN_DIM)}
+        # 64-D token proprio state, aggregated by rollout into observation.state (last token).
+        self.observation_ft = {f"{TOKEN_STATE_PREFIX}.{i}.pos": float for i in range(TOKEN_DIM)}
+
+        self.reset()
+        logger.info("SonicWholeBodyController initialized")
+
+    def reset(self) -> None:
+        """Reset internal state for a new episode: held token and 10-frame history buffers."""
+        self.last_action_mj = np.zeros(29, np.float32)
```
Comment (CarolinePascal) [inline src/lerobot/robots/unitree_g1/controllers/sonic_whole_body.py]: Nit pick: would be nice to have a global var for the 29 DOFs

File: src/lerobot/robots/unitree_g1/controllers/sonic_whole_body.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        self.h_quat = [np.array([1, 0, 0, 0], np.float32)] * 10
+        self._last_token = None  # neutral token is re-seeded on the first tick
+
+    def observation_state(self) -> dict[str, float]:
+        """Echo the last decoded token as ``observation.state`` so a token-output VLA closes
+        the loop on its own previous token."""
+        token = self._last_token if self._last_token is not None else np.zeros(TOKEN_DIM, dtype=np.float32)
+        return {f"{TOKEN_STATE_PREFIX}.{i}.pos": float(v) for i, v in enumerate(token)}
+
+    def run_step(self, action: dict, lowstate) -> dict:
```
Comment (CarolinePascal) [inline src/lerobot/robots/unitree_g1/controllers/sonic_whole_body.py:152]: Could you add docstrings here ?

File: src/lerobot/robots/unitree_g1/controllers/sonic_whole_body.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    missing = [k for k in required if k not in metadata]
+    if missing:
+        raise ValueError(f"ONNX model must contain {list(required)} in metadata (missing {missing})")
+
+    arr = {k: np.array(json.loads(metadata[k]), dtype=np.float32) for k in required}
+    logger.info(f"Loaded SONIC deploy constants from ONNX ({len(arr['kp'])} joints)")
+    return decoder, arr["kp"], arr["kd"], arr["default_angles"], arr["action_scale"], arr["neutral_token"]
+
+
+class SonicWholeBodyController:
```
Comment (CarolinePascal) [inline src/lerobot/robots/unitree_g1/controllers/sonic_whole_body.py]: Why doesn't this inherit `LocomotionController` ? Also, I feel like `LocomotionController could be extended/completed !

File: src/lerobot/robots/unitree_g1/config_unitree_g1.py [hunk 7]
```diff
@@ -68,6 +68,6 @@ class UnitreeG1Config(RobotConfig):
     # Compensates for gravity on the unitree's arms using the arm ik solver
     gravity_compensation: bool = False
 
-    # Lower-body controller class name, e.g. "GrootLocomotionController" or
-    # "HolosomaLocomotionController". None disables it.
+    # Controller class name, e.g. GrootLocomotionController / HolosomaLocomotionController /
+    # SonicWholeBodyController. None disables it.
     controller: str | None = None
```
Comment (CarolinePascal) [inline src/lerobot/robots/unitree_g1/config_unitree_g1.py:73]: Ideally, we should create a `LocomotionControllerConfig` to be used here.

**Review by CarolinePascal — COMMENTED** [review]
Looks good ! Left a few reviews :) Could you also tackle Claude's review ?

File: src/lerobot/robots/unitree_g1/controllers/sonic_whole_body.py (hunk 4 above)
Comment (nepyope) [reply src/lerobot/robots/unitree_g1/controllers/sonic_whole_body.py]: Good idea to unify all the controller behaviors. Starting with this, i moved NUM_MOTORS to g1_utils.py, each controller now imports it.

File: src/lerobot/robots/unitree_g1/controllers/sonic_whole_body.py (hunk 5 above)
Comment (nepyope) [reply src/lerobot/robots/unitree_g1/controllers/sonic_whole_body.py:152]: added below

File: src/lerobot/robots/unitree_g1/controllers/sonic_whole_body.py (hunk 6 above)
Comment (nepyope) [reply src/lerobot/robots/unitree_g1/controllers/sonic_whole_body.py]: SonicWholeBodyController inherits it now like the others; I renamed it to
RobotController which is more general as SONIC drives all joints unlike the other two.
`make_locomotion_controller` is now `make_robot_controller`.
inside it, i declare control_dt, run_step() and reset() and the docstring
now spells out the optional hooks (kp/kd, default_angles, action_ft/observation_ft and observation_state()).

File: src/lerobot/robots/unitree_g1/config_unitree_g1.py (hunk 7 above)
Comment (nepyope) [reply src/lerobot/robots/unitree_g1/config_unitree_g1.py:73]: Better suited as a follow-up imo

**Comment by nepyope** [comment]
Added tests

**Comment by CarolinePascal** [comment]
@claude can you review again and make sure all the comments were addressed ?

**Review by CarolinePascal — APPROVED** [review]
LGTM !
